pilot/pkg/model/jwks_resolver.go

```diff
@@ -67,7 +67,7 @@ const (
 	// How many times should we retry the failed network fetch on main flow. The main flow
 	// means it's called when Pilot is pushing configs. Do not retry to make sure not to block Pilot
 	// too long.
-	networkFetchRetryCountOnMainFlow = 0
+	networkFetchRetryCountOnMainFlow = 5
```

Comment:
This is on the listener generation path, not sure if `5` is too high for it, probably set it to a smaller value `1` or `2`? 

I think @howardjohn has better insights on this.

Reply:
I have tried 1, 2, 3 and it still fails with that.

Reply:
If we don't want to increase the retries, other option is to reduce the refresh interval so that we can try to fetch the public key in background.

Reply:
We should not have any retries on the push path - I don't even think we should have any calls at all. We should do it all async in the background. When a user creates a new ReqAuth, we do not push it immediately but schedule a background job to fetch the jwks. Then once it fetches we push.

This is the pattern in https://github.com/istio/istio/blob/e734af4be2312475bea8c9367abb4138cb44ae16/pilot/pkg/model/network.go#L396.

Reply:
The other option is as john suggested, completely refactor the code, remove the call on the push flow and fetch the jwksUri asynchronously. I think this probably does not need much code changes since we already have the JwksResolver here (it is already started as a separate job for the jwks refresh, just need to update it to handle the push path use case as well): https://github.com/istio/istio/blob/cd004f1ba7f9bab3951824912af4119074ec9a00/pilot/pkg/model/jwks_resolver.go#L154

Reply:
the time for retry in async process is so high - https://github.com/istio/istio/blob/cd004f1ba7f9bab3951824912af4119074ec9a00/pilot/pkg/model/jwks_resolver.go#L58
By that time out test ends, is it okay to decrease this to some smaller time ?

Reply:
@yangminzhu correct me if I am wrong do we not have this logic already ? We have a async function which hits the jwks uri continuously after a fixed interval of time and add the key to the cache.

Reply:
The idea here is to do some refactor, let the main listener push flow to trigger the refresh() job to fetch the jwksUri asynchronously, once the refresh() job fetched the jwks, it needs to trigger an extra config push. The refresh() can retry with a much smaller interval if it failed, there is a separate refresh interval for the refresher() job to rotate jwks keys (that are already fetched successfully).

Reply:
@howardjohn By the way, if we remove the fetch from the main flow completely, it seems we will always trigger a second config push right after the first one (when the jwksUri fetching is done), is this kind of consecutive config push desired?

Reply:
Well it would only be when there is a new JWKS added. That must be extremely rare. If the code is smart enough it will not push on the first creation and only when the jwks is resolved

Reply:
we can modify the push, if it doesn't have any pubkey we can stop that. Right ?